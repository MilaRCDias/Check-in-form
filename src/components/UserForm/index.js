import React, {useState} from 'react';
import PropTypes from "prop-types";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import { makeStyles, createStyles } from "@material-ui/core/styles";
import {
  TextField,
  Box,
  Button,
  Checkbox,
  FormControlLabel,
  InputLabel,
  FormControl,
  Select,
  MenuItem,
  Typography,
} from "@material-ui/core";

const useStyles = makeStyles(() =>
  createStyles({
    error: {
      color: "red",
      fontStyle: "italic",
    },
  })
);


/**
 * User Form Component
 * Step 2 to user insert all informations
 */
const UserForm = ({ formValues, currentStep, handleStep, setStep }) => {
  const style = useStyles();
  const [userNationality, setUserNationality] = useState("");
  const userName =
    formValues.lastName?.charAt(0).toUpperCase() +
    formValues.lastName?.slice(1);

  //TODO get from api
  const nationalityList = [
    "Portuguese",
    "Grand Britain",
    "Germany",
    "Austria",
    "Belgium",
    "France",
    "Greece",
    "Spain",
    "Italia",
  ];

  /** */
  const initialValue = {
    firstName: "",
    lastName: formValues.lastName || null,
    nationality: "",
    email: "",
    phoneNumber: "",
    passportNumber: "",
    city: "",
    country: "",
    address: "",
    birthdate: "",
    birthplace: "",
    passportExpireDate: "",
    passportDateIssue: "",
    passportCityIssue: "",
    passportCountryIssue: "",
    terms: "",
  };
  /** */
  const validation = Yup.object().shape({
    firstName: Yup.string()
      .min(2, "Too Short!")
      .max(20, "Too Long!")
      .required("Firstname is required"),
    lastName: Yup.string()
      .min(2, "Too Short!")
      .max(50, "Too Long!")
      .required("Lastname is required"),
    nationality: Yup.string().required("Select a nationality"),
    email: Yup.string().email().required("Email is required"),
    phoneNumber: Yup.string()
      .matches(
        /^(\+?\d{0,4})?\s?-?\s?(\(?\d{3}\)?)\s?-?\s?(\(?\d{3}\)?)\s?-?\s?(\(?\d{4}\)?)?$/,
        "Invalid phone"
      )
      .required("Phone number is required"),
    passportNumber: Yup.string()
      .min(2, "Too Short!")
      .required("enter a valid passport number")
      .test({
        test: function (passportNumber) {
          const isValid = /[!@#$%^&*(),.?":{}|<>]/.test(passportNumber);
          return !isValid;
        },
      }),
    city: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Austria", "Belgium", "France"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    country: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Austria", "Belgium", "France"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    address: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Belgium", "Spain"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    birthdate: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Belgium", "France"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    birthplace: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["France"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),

    passportExpireDate: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Austria", "Greece"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    passportDateIssue: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Greece"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    passportCityIssue: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Greece"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    passportCountryIssue: Yup.string()
      .nullable()
      .when("nationality", {
        is: () => ["Greece"].includes(userNationality),
        then: Yup.string().nullable().required("required"),
      }),
    terms: Yup.boolean().required("Terms and conditions are required"),
  });

  /** */
  const onSubmitForm = (values) => {
    console.log(values);
    handleStep(values);
    setStep(3);
  };

  return (
    <div>
      {currentStep === 2 ? (
        <div>
          {" "}
          <Box mt={8} mx={4}>
            <Typography
              variant="h4"
              component="h1"
              align="left"
              color="textSecondary"
            >
              {" "}
              Hi, {userName}{" "}
            </Typography>
            <Typography variant="body1" align="left" color="textSecondary">
              Please fill your informations
            </Typography>
          </Box>
          <Formik
            initialValues={initialValue}
            validationSchema={validation}
            onSubmit={onSubmitForm}
          >
            {({ errors, touched, handleChange, isSubmitting }) => {
              const handleInfoChange = (e) => {
                handleChange(e);
                setUserNationality(e.target.value);
              };

              return (
                <Form>
                  <Box m={4}>
                    <TextField
                      id="firstName"
                      name="firstName"
                      label="first name"
                      fullWidth
                      variant="outlined"
                      onChange={handleChange}
                      error={!!errors.firstName}
                    />
                    {errors.firstName && touched.firstName ? (
                      <div className={style.error}>{errors.firstName}</div>
                    ) : null}
                  </Box>
                  <Box m={4}>
                    <TextField
                      id="lastName"
                      name="lastName"
                      label="Last Name"
                      disabled
                      fullWidth
                      variant="outlined"
                      value={formValues.lastName}
                      onChange={handleChange}
                      error={!!errors.lastName}
                    />
                    {errors.lastName && touched.lastName ? (
                      <div className={style.error}>{errors.lastName}</div>
                    ) : null}
                  </Box>
                  <Box m={4}>
                    <FormControl
                      fullWidth
                      variant="outlined"
                      error={!!errors.nationality}
                    >
                      <InputLabel id="nationality">Nationality</InputLabel>
                      <Select
                        labelId="nationality"
                        id="nationality"
                        name="nationality"
                        label="Nationality"
                        value={userNationality}
                        onChange={handleInfoChange}
                      >
                        <MenuItem value="" disabled>
                          Choose
                        </MenuItem>
                        {nationalityList.map((option) => (
                          <MenuItem key={option} value={option}>
                            {option}
                          </MenuItem>
                        ))}
                      </Select>
                      {errors.nationality && touched.nationality ? (
                        <div className={style.error}>{errors.nationality}</div>
                      ) : null}
                    </FormControl>
                  </Box>
                  {["Austria", "France"].includes(userNationality) ? (
                    <Box>
                      <Box m={4}>
                        <TextField
                          id="city"
                          name="city"
                          label="City of residence"
                          fullWidth
                          variant="outlined"
                          onChange={handleChange}
                          error={!!errors.city}
                        />
                        {errors.city && touched.city ? (
                          <div className={style.error}>{errors.city}</div>
                        ) : null}
                      </Box>
                      <Box m={4}>
                        <TextField
                          id="country"
                          name="country"
                          label="Country of residence"
                          fullWidth
                          variant="outlined"
                          onChange={handleChange}
                          error={!!errors.country}
                        />
                        {errors.country && touched.country ? (
                          <div className={style.error}>{errors.country}</div>
                        ) : null}
                      </Box>
                    </Box>
                  ) : null}
                  {["Belgium", "Spain"].includes(userNationality) ? (
                    <Box m={4}>
                      <TextField
                        id="address"
                        label="Address"
                        name="address"
                        fullWidth
                        variant="outlined"
                        onChange={handleChange}
                        error={!!errors.address}
                      />
                      {errors.address && touched.address ? (
                        <div className={style.error}>{errors.address}</div>
                      ) : null}
                    </Box>
                  ) : null}

                  {["Belgium", "France"].includes(userNationality) ? (
                    <Box m={4}>
                      <TextField
                        id="birthdate"
                        label="Date of birth"
                        name="birthdate"
                        type="date"
                        fullWidth
                        variant="outlined"
                        InputLabelProps={{
                          shrink: true,
                        }}
                        onChange={handleChange}
                        error={!!errors.birthdate}
                      />
                      {errors.birthdate && touched.birthdate ? (
                        <div className={style.error}>{errors.birthdate}</div>
                      ) : null}
                    </Box>
                  ) : null}
                  {userNationality === "France" ? (
                    <Box m={4}>
                      <TextField
                        id="birthplace"
                        name="birthplace"
                        label="Place of birth"
                        fullWidth
                        variant="outlined"
                        onChange={handleChange}
                        error={!!errors.birthplace}
                      />
                      {errors.birthplace && touched.birthplace ? (
                        <div className={style.error}>{errors.birthplace}</div>
                      ) : null}
                    </Box>
                  ) : null}
                  <Box m={4}>
                    <TextField
                      id="email"
                      name="email"
                      label="email"
                      fullWidth
                      variant="outlined"
                      onChange={handleChange}
                      error={!!errors.email}
                    />
                    {errors.email && touched.email ? (
                      <div className={style.error}>{errors.email}</div>
                    ) : null}
                  </Box>
                  <Box m={4}>
                    <TextField
                      id="phoneNumber"
                      name="phoneNumber"
                      label="Phone Number"
                      fullWidth
                      variant="outlined"
                      onChange={handleChange}
                      error={!!errors.phoneNumber}
                    />
                    {errors.phoneNumber && touched.phoneNumber ? (
                      <div className={style.error}>{errors.phoneNumber}</div>
                    ) : null}
                  </Box>
                  <Box m={4}>
                    <TextField
                      id="passportNumber"
                      name="passportNumber"
                      label="Passport Number"
                      fullWidth
                      variant="outlined"
                      onChange={handleChange}
                      error={!!errors.passportNumber}
                    />
                    {errors.passportNumber && touched.passportNumber ? (
                      <div className={style.error}>{errors.passportNumber}</div>
                    ) : null}
                  </Box>
                  {["Austria", "Greece"].includes(userNationality) ? (
                    <Box m={4}>
                      <TextField
                        id="passportExpireDate"
                        name="passportExpireDate"
                        label="Passport expire date"
                        type="date"
                        fullWidth
                        variant="outlined"
                        onChange={handleChange}
                        InputLabelProps={{
                          shrink: true,
                        }}
                        error={!!errors.passportExpireDate}
                      />
                      {errors.passportExpireDate &&
                      touched.passportExpireDate ? (
                        <div className={style.error}>
                          {errors.passportExpireDate}
                        </div>
                      ) : null}
                    </Box>
                  ) : null}
                  {userNationality === "Greece" ? (
                    <Box>
                      <Box m={4}>
                        <TextField
                          id="passportDateIssue"
                          name="passportDateIssue"
                          label="Passport date of issue"
                          type="date"
                          fullWidth
                          variant="outlined"
                          InputLabelProps={{
                            shrink: true,
                          }}
                          onChange={handleChange}
                          error={!!errors.passportDateIssue}
                        />
                        {errors.passportDateIssue &&
                        touched.passportDateIssue ? (
                          <div className={style.error}>
                            {errors.passportDateIssue}
                          </div>
                        ) : null}
                      </Box>
                      <Box m={4}>
                        <TextField
                          id="passportCityIssue"
                          name="passportCityIssue"
                          label="Passport city of issue"
                          fullWidth
                          variant="outlined"
                          onChange={handleChange}
                          error={!!errors.passportCityIssue}
                        />
                        {errors.passportCityIssue &&
                        touched.passportCityIssue ? (
                          <div className={style.error}>
                            {errors.passportCityIssue}
                          </div>
                        ) : null}
                      </Box>
                      <Box m={4}>
                        <TextField
                          id="passportCountryIssue"
                          name="passportCountryIssue"
                          label="Passport country of issue"
                          fullWidth
                          variant="outlined"
                          onChange={handleChange}
                          error={!!errors.passportCountryIssue}
                        />
                        {errors.passportCountryIssue &&
                        touched.passportCountryIssue ? (
                          <div className={style.error}>
                            {errors.passportCountryIssue}
                          </div>
                        ) : null}
                      </Box>
                    </Box>
                  ) : null}
                  <Box m={4}>
                    <FormControlLabel
                      control={
                        <Checkbox
                          //checked={termsChecked}
                          onChange={handleChange}
                          name="terms"
                          color="primary"
                          error={errors.terms}
                        />
                      }
                      label="Terms & conditions"
                    />
                    {errors.terms && touched.terms ? (
                      <div className={style.error}>{errors.terms}</div>
                    ) : null}
                  </Box>
                  <Button
                    disabled={isSubmitting}
                    type="submit"
                    variant="contained"
                    color="primary"
                  >
                    Continue
                  </Button>
                </Form>
              );
            }}
          </Formik>
        </div>
      ) : null}
    </div>
  );
};

 UserForm.propTypes={
     /** the number of the form step */
  currentStep: PropTypes.number.isRequired,
  /** handle submit form */
  handleStep: PropTypes.func.isRequired,
  /** object with form information */
  formValues: PropTypes.shape({}).isRequired,
  /** set next step */
  setStep: PropTypes.func.isRequired,
}



export default UserForm;