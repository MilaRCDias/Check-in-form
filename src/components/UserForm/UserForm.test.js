import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import testConfig from "../../../testConfig";
import { mount } from "enzyme";
import UserForm from "./index";

describe("the welcome checkin step", () => {
  let shallow;
  const props = {
    currentStep: 1,
    handleStep: () => {},
    formValues: {},
    setStep: () => {},
    nationalityList: [],
  };
  beforeAll(() => {
    shallow = createShallow();
  });

  it("should capture first name", () => {
    const credentials = { firstName: "first name" };

    const wrapper = mount(shallow(<UserForm {...props} />).get(0));
    const input = wrapper.find("#firstName");

    input.value = credentials.firstName;
    expect(input.value).toEqual("first name");
  });

  it("should capture passport number", () => {
    const credentials = { passportNumber: "ee5432" };

    const wrapper = mount(shallow(<UserForm {...props} />).get(0));
    const input = wrapper.find("#passportNumber");

    input.value = credentials.passportNumber;
    expect(input.value).toEqual("ee5432");
  });

 it("should capture nationality", () => {
   const credentials = { nationality: "Austria" };

   const wrapper = mount(shallow(<UserForm {...props} />).get(0));
   const input = wrapper.find("#nationality");

   input.value = credentials.nationality;
   expect(input.value).toEqual("Austria");
 });

  it("should match the snapshot", () => {
    const wrapper = mount(shallow(<UserForm {...props} />).get(0));

    expect(wrapper.html()).toMatchSnapshot();
  });
});
