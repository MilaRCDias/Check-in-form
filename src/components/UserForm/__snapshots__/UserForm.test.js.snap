// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the welcome checkin step should match the snapshot 1`] = `"<div><div class=\\"MuiBox-root MuiBox-root-34\\"><h1 class=\\"MuiTypography-root MuiTypography-h4 MuiTypography-colorTextSecondary MuiTypography-alignLeft\\"> Hi,  </h1><p class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-alignLeft\\">Please fill your informations</p></div><form action=\\"#\\"><div class=\\"MuiBox-root MuiBox-root-35\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"firstName\\" id=\\"firstName-label\\">first name</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"firstName\\" name=\\"firstName\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-6\\"><span>first name</span></legend></fieldset></div></div></div><div class=\\"MuiBox-root MuiBox-root-36\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled\\" data-shrink=\\"false\\" for=\\"lastName\\" id=\\"lastName-label\\">Last Name</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" disabled=\\"\\" id=\\"lastName\\" name=\\"lastName\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-6\\"><span>Last Name</span></legend></fieldset></div></div></div><div class=\\"MuiBox-root MuiBox-root-37\\"><div class=\\"MuiFormControl-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined\\" data-shrink=\\"false\\" id=\\"nationality\\">Nationality</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl\\"><div class=\\"MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input\\" tabindex=\\"0\\" role=\\"button\\" aria-haspopup=\\"listbox\\" aria-labelledby=\\"nationality nationality\\" id=\\"nationality\\"><span>â€‹</span></div><input name=\\"nationality\\" aria-hidden=\\"true\\" tabindex=\\"-1\\" class=\\"MuiSelect-nativeInput\\" value=\\"\\"><svg class=\\"MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-6\\"><span>Nationality</span></legend></fieldset></div></div></div><div class=\\"MuiBox-root MuiBox-root-38\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"email\\" id=\\"email-label\\">email</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"email\\" name=\\"email\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-6\\"><span>email</span></legend></fieldset></div></div></div><div class=\\"MuiBox-root MuiBox-root-39\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"phoneNumber\\" id=\\"phoneNumber-label\\">Phone Number</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"phoneNumber\\" name=\\"phoneNumber\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-6\\"><span>Phone Number</span></legend></fieldset></div></div></div><div class=\\"MuiBox-root MuiBox-root-40\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"passportNumber\\" id=\\"passportNumber-label\\">Passport Number</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"passportNumber\\" name=\\"passportNumber\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-6\\"><span>Passport Number</span></legend></fieldset></div></div></div><div class=\\"MuiBox-root MuiBox-root-41\\"><label class=\\"MuiFormControlLabel-root\\"><span class=\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-14 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary\\" aria-disabled=\\"false\\"><span class=\\"MuiIconButton-label\\"><input class=\\"PrivateSwitchBase-input-17\\" name=\\"terms\\" type=\\"checkbox\\" data-indeterminate=\\"false\\" value=\\"\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></span><span class=\\"MuiTypography-root MuiFormControlLabel-label MuiTypography-body1\\">Terms &amp; conditions</span></label></div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"submit\\"><span class=\\"MuiButton-label\\">Continue</span><span class=\\"MuiTouchRipple-root\\"></span></button></form></div>"`;
